File: database/database-annotation-processor/src/test/java/ru/tinkoff/kora/database/common/annotation/processor/AbstractClassTest.java
package ru.tinkoff.kora.database.common.annotation.processor;

import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import ru.tinkoff.kora.database.annotation.processor.RepositoryAnnotationProcessor;
import ru.tinkoff.kora.database.common.annotation.processor.jdbc.AbstractJdbcRepositoryTest;
import ru.tinkoff.kora.kora.app.annotation.processor.KoraAppProcessor;

import java.sql.SQLException;
import java.util.List;

public class AbstractClassTest extends AbstractJdbcRepositoryTest {
    @Test
    public void testAbstractClassRepository() throws SQLException {
        var repository = compile("test", List.of(executor), """
            @Repository
            public abstract class TestRepository implements JdbcRepository {
                private final String field;
                        
                public TestRepository(@Nullable String field) {
                    this.field = field;
                }
                        
                @Query("INSERT INTO table(value) VALUES (:value)")
                public abstract void abstractMethod(String value);
                        
                public void nonAbstractMethod() {
                        
                }
            }
            """);

        repository.invoke("abstractMethod", "some-value");

        Mockito.verify(executor.preparedStatement).setString(1, "some-value");
    }

    @Test
    public void testAbstractClassRepositoryExtension() throws SQLException {
        compile(List.of(new RepositoryAnnotationProcessor(), new KoraAppProcessor()), """
            @Repository
            public abstract class TestRepository implements JdbcRepository {
                @Query("INSERT INTO table(value) VALUES (:value)")
                public abstract void abstractMethod(String value);
                        
                public void nonAbstractMethod() {
                        
                }
            }
            """, """
            @KoraApp
            public interface TestApp {
                default ru.tinkoff.kora.database.jdbc.JdbcConnectionFactory factory() { return null; }
                default String someString() { return null; }
                
                @Root
                default Integer someRoot(TestRepository repository) { return 1; }
            }
            """);
        compileResult.assertSuccess();

    }
}

File: database/database-symbol-processor/src/test/kotlin/ru/tinkoff/kora/database/symbol/processor/AbstractClassTest.kt
package ru.tinkoff.kora.database.symbol.processor

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import ru.tinkoff.kora.database.symbol.processor.jdbc.AbstractJdbcRepositoryTest
import ru.tinkoff.kora.kora.app.ksp.KoraAppProcessorProvider

class AbstractClassTest : AbstractJdbcRepositoryTest() {

    @Test
    fun testAbstractClassRepository() {
        compile(listOf(RepositorySymbolProcessorProvider()), """
            @Repository
            abstract class AbstractClassRepository(private val field: String?) : JdbcRepository {
                @Query("INSERT INTO table(value) VALUES (:value)")
                abstract fun abstractMethod(value: String?)
                fun nonAbstractMethod() {}
            }
        """.trimIndent())

        compileResult.assertSuccess()
        assertThat(compileResult.loadClass("\$AbstractClassRepository_Impl")).isFinal
    }

    @Test
    fun testAbstractClassRepositoryExtension() {
        compile(listOf(RepositorySymbolProcessorProvider(), KoraAppProcessorProvider()), """
            @Repository
            abstract class AbstractClassRepository : JdbcRepository {
                @Query("INSERT INTO table(value) VALUES (:value)")
                abstract fun abstractMethod(value: String?)
            }
        """.trimIndent(), """
            @KoraApp
            interface TestApp {
                fun factory() : ru.tinkoff.kora.database.jdbc.JdbcConnectionFactory {
                  return org.mockito.Mockito.mock(ru.tinkoff.kora.database.jdbc.JdbcConnectionFactory::class.java)
                }
                
                @Root
                fun root(repository: AbstractClassRepository) : String = "test"
            }
        """.trimIndent())

        compileResult.assertSuccess()
        assertThat(compileResult.loadClass("\$AbstractClassRepository_Impl")).isFinal
    }
}
